### PR TITLE
Infrastructure

### PR DESCRIPTION
**Pull request by bwagner5**
Issue #, if available:
N/A

Description of changes:
 - Adds a CDKv2 test infra application which installs
   - KIT
   - Tekton
   - AWS LB Controller
   - Flux v2
   - EBS CSI Driver
   - Karpenter
   - Crossplane

The cdk app bootstraps two flux configurations, one watching the KIT repo path, and the other watching a configurable application repo that you are using the KIT infra to test. 

For example, Karpenter uses the KIT infra cdk app to stand-up the test infrastructure and then a Flux config monitors a path in the Karpenter repo for other components to be installed like the Provisioner spec and test specs. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: testbed/bin/infra.ts
Comment (ellistarn) [inline testbed/bin/infra.ts:7]: re: messaging/communication, WDYT about KITInfrastructure? 

`github.com/kit/infrastructure/...`

`new KITInfrastructure`

File: testbed/README.md
Comment (ellistarn) [inline testbed/README.md:38]: Can we hardcode these, such that every KITInfrastructure just magically comes with this? I can see an argument for overriding `KITRepoURL`

File: testbed/README.md
Comment (ellistarn) [inline testbed/README.md:41]: Consider `git-ref` for more flexibility

File: testbed/README.md
Comment (bwagner5) [reply testbed/README.md:38]: I think they should all be able to be overridden.  But I can add defaults, so that you'll get these by default. sound good?

File: testbed/README.md
Comment (ellistarn) [inline testbed/README.md:40]: Ideally we'd leave this out of the KIT repo and replace with something like ${SYNC_REPOSITORY}

File: testbed/README.md
Comment (ellistarn) [inline testbed/README.md:44]: optional: I kind of wish we just used the namespace's default service account.

File: testbed/README.md
Comment (bwagner5) [reply testbed/README.md:40]: yeah we could do that. I thought this might be more helpful though to show an example repo using KIT .

File: testbed/lib/infra-stack.ts
Comment (ellistarn) [inline testbed/lib/infra-stack.ts:117]: Do you foresee multiple test namespaces? If we expect all tests within a single cluster to live in a single namespace, it might be kind of nice to just standardize on a well known namespace that's consistent across projects e.g. `test-suites`

File: testbed/lib/infra-stack.ts
Comment (ellistarn) [inline testbed/lib/infra-stack.ts:159]: Consider hardcoding things like this, unless you expect it to change.

File: testbed/lib/infra-stack.ts
Comment (ellistarn) [inline testbed/lib/infra-stack.ts:171]: likewise.

File: testbed/test/infra.test.ts
Comment (ellistarn) [inline testbed/test/infra.test.ts:1]: Is this doing anything?

File: testbed/lib/addons/kit.ts
Comment (ellistarn) [inline testbed/lib/addons/kit.ts:33]: dd

File: testbed/lib/addons/kit.ts
Comment (ellistarn) [inline testbed/lib/addons/kit.ts:35]: dd

File: testbed/lib/addons/karpenter.ts
Comment (ellistarn) [inline testbed/lib/addons/karpenter.ts:29]: If we hardcode this policy, we should probably hardcode the karpenter version inside this module, since these may drift.

File: testbed/lib/infra-stack.ts
Comment (bwagner5) [reply testbed/lib/infra-stack.ts:159]: constructs are generic and a lot of people run these in different namespaces, but I'd rather leave it like this. I think this provides a nice at-a-glance view of the infra being deployed and the namespace.

**Review by ellistarn — COMMENTED** [review]
Awesome! A few minor things

File: testbed/README.md
Comment (bwagner5) [reply testbed/README.md:41]: I don't think git-ref adds a lot of value in the case of flux syncing repos. I think you always want your test infra looking at head of a branch.  Adding git-ref is possible but it's not a drop-in so would rather leave it out until/if we need it.

File: testbed/lib/infra-stack.ts
Comment (bwagner5) [reply testbed/lib/infra-stack.ts:171]: yep, but.... to my knowledge that's arbitrary just because it's easiest. Seems better to run in separate namespaces, if for nothing else, ease of seeing an application resources.

File: testbed/README.md
Comment (ellistarn) [reply testbed/README.md:41]: I want my own personal KITInfrastructure pointing at github.com/ellistarn/karpenter so I can run presubmits before I merge.

File: testbed/README.md
Comment (ellistarn) [reply testbed/README.md:38]: SGTM!

**Review by ellistarn — COMMENTED** [review]
All good by me -- get a review from @prateekgogia

File: testbed/README.md
Comment (bwagner5) [reply testbed/README.md:41]: You can do that with a branch, you just change the TestRepoURL to your fork and the branch to your feature branch

File: testbed/README.md
Comment (bwagner5) [reply testbed/README.md:44]: I agree, and I originally tried to do that. Unfortunately it was more difficult with CDK since the default namespace gets created automatically, so if you try to create it, you get a duplicate error and it's hard to reference it as a dependency in the cdk app if you just assume it's there.

**Review by prateekgogia — COMMENTED** [review]
Great work!

Overall looks great. Can we also include the default [Karpenter provisioner used by KIT](https://github.com/awslabs/kubernetes-iteration-toolkit/blob/main/substrate/pkg/controller/substrate/cluster/addons/karpenter.go#L36) to provision nodes for guest clusters in the `k8s-config/clusters/kit-infrastructure` which can be created by Flux.

File: infrastructure/lib/addons/crossplane.ts
Comment (bwagner5) [reply infrastructure/lib/addons/crossplane.ts:109]: They'll likely be needed later. I didn't get as far a long on crossplane as I'd like, but wanted to get this in earlier.

File: infrastructure/lib/addons/karpenter.ts
Comment (bwagner5) [reply infrastructure/lib/addons/karpenter.ts:32]: There's not a policy document, there is a cloudformation template though. It's on my todo list to change this out to a dynamically updating policy. But wanted to get this MVP in that works. I think exposing the policy within Karpenter like AWS LBC does is a better experience than the CFT we have today anyways, so wanted to work to change all of that which will require a bit more effort, but coming soon!

File: infrastructure/lib/addons/karpenter.ts
Comment (bwagner5) [reply infrastructure/lib/addons/karpenter.ts:67]: I'd like to take a follow-on and investigate approaches to making it work with the latest version of Karpenter. I think we can make it work.

File: infrastructure/lib/addons/kit.ts
Comment (bwagner5) [reply infrastructure/lib/addons/kit.ts:36]: Yeah, I think that's a good idea! When I work on adding one to Karpenter, maybe I can copy over the same approach for KIT!

**Comment by hakuna-matatah** [comment]
@ellistarn Isn't this approach using CFN in the backend similar to what we have done before ?  but just with the updated/stable cdkV2 version  ? Or am i missing anything here ?

File: infrastructure/README.md
Comment (bwagner5) [reply infrastructure/README.md:38]: > @ellistarn @bwagner5 Isn't this approach using CFN in the backend similar to what we have done before ? but just with the updated/stable cdkV2 version ? Or am i missing anything here ?

@hakuna-matatah and I had a side discussion, but want to leave a summary here. This is very similar to the testbed approach before. The trade-off is in maintaining the infrastructure and having the operational benefits of EKS rather than using kitctl right now.

**Comment by bwagner5** [comment]
> @ellistarn @bwagner5 Isn't this approach using CFN in the backend similar to what we have done before ? but just with the updated/stable cdkV2 version ? Or am i missing anything here ?

@hakuna-matatah and I had a side discussion, but want to leave a summary here. This is very similar to the testbed approach before. The trade-off is in maintaining the infrastructure and having the operational benefits of EKS rather than using kitctl right now.

**Review by prateekgogia — APPROVED** [review]
LGTM!

**Pull request merged**
